packages/permissions/src/types.ts

```diff
@@ -0,0 +1,111 @@
+export type PERMISSION_NAMES =
+  | PermissionName
+  | "background-sync"
+  | "compute-pressure"
+  | "geolocation"
+  | "local-fonts"
+  | "microphone"
+  | "camera"
+  | "notifications"
+  | "payment-handler"
+  | "push"
+  | "screen-wake-lock"
+  | "accelerometer"
+  | "gyroscope"
+  | "magnetometer"
+  | "ambient-light-sensor"
+  | "storage-access"
+  | "top-level-storage-access"
+  | "persistent-storage"
+  | "midi"
+  | "window-management"
+  | "accessibility-events"
+  | "bluetooth"
+  | "clipboard-read"
+  | "clipboard-write";
+
+export type PermissionOption =
+  | {
+      name: PERMISSION_NAMES;
+    }
+  | {
+      name: "midi";
+      sysex?: boolean;
+    }
+  | {
+      name: "push";
+      userVisibleOnly?: boolean;
+    };
+
+type PermissionGranted = {
+  state: "granted" | "prompt";
+  status: PermissionStatus;
+  message: string;
+};
+
+type PermissionError = {
+  state: "unsupported" | "invalid";
+  message: string;
+};
+
+type PermissionDenied = {
+  state: "denied";
+  status: PermissionStatus;
+  message: string;
+};
+
+type PermissionState = {
+  state: "granted" | "prompt" | "denied";
+  status: PermissionStatus;
+  message: string;
+};
+
+export type PermissionResponse<
+  Status extends "granted" | "denied" | "error" | void = void,
+> = Status extends "granted"
+  ? PermissionGranted
+  : Status extends "denied"
+    ? PermissionDenied
+    : Status extends "error"
+      ? PermissionError
+      : PermissionState;
```

Comment:
The type alias `PermissionResponse` is using an extended conditional type which can be simplified to improve readability and maintainability.

```diff
- export type PermissionResponse<
-   Status extends "granted" | "denied" | "error" | void = void,
- > = Status extends "granted"
-   ? PermissionGranted
-   : Status extends "denied"
-     ? PermissionDenied
-     : Status extends "error"
-       ? PermissionError
-       : PermissionState;
+ export type PermissionResponse<Status = void> =
+   Status extends 'granted' ? PermissionGranted :
+   Status extends 'denied' ? PermissionDenied :
+   Status extends 'error' ? PermissionError :
+   PermissionState;
```